qiskit/providers/backend.py

```diff
@@ -229,3 +240,435 @@ class can handle either situation.
             Job: The job object for the run
         """
         pass
+
+
+class QubitProperties:
+    """A representation of the properties of a qubit on a backend."""
+
+    __slots = ("t1", "t2", "frequency", "properties")
+
+    def __init__(self, t1=None, t2=None, frequency=None, properties=None):
+        """Create a new ``QubitProperties`` object
+
+        Args:
+            t1: The T1 time for a qubit in seconds
+            t2: The T2 time for a qubit in seconds
+            frequency: The frequency of a qubit in Hz
+            properties: A free form dictionary of additional properties the
```

Comment:
anharmonicity goes here? (seems bit awkward since only anharmonicity has less accessibility) Can we turn this into abstract and let the provider override the attributes? In our case this can be provided as `TransmonQubitProperties` for example.

Reply:
Related with the above question, since all attributes are Optional (allowing None), what's the difference between non-additional "optional" attributes (e.g. `t1`) and "additional" attributes `properties`? (It seems just because V1 had dedicated access functions for the non-additional attributes.)
I thought qubit properties which may be used in transpilers should live in `target`, however, it's difficult to anticipate which attributes will be used in the future, noise-aware passes might want to know all of the qubit properties. I have no good idea, but just want to bring it up.

Reply:
> anharmonicity goes here? (seems bit awkward since only anharmonicity has less accessibility) Can we turn this into abstract and let the provider override the attributes? In our case this can be provided as `TransmonQubitProperties` with anharmonicity for example.

I'm not sure if an abstract is needed, a provider could subclass this easily and add additional properties that way. Maybe I should remove the `properties` attributes (both here and in `InstructionProperties`) and document that adding additional properties is valid and expected and the mechanism to do that is to subclass and add new attributes? I don't have a super strong preference either way.

> Related with the above question, since all attributes are Optional (allowing None), what's the difference between non-additional "optional" attributes (e.g. `t1`) and "additional" attributes `properties`? (It seems just because V1 had dedicated access functions for the non-additional attributes.) I thought qubit properties which may be used in transpilers should live in `target`, however, it's difficult to anticipate which attributes will be used in the future, noise-aware passes might want to know all of the qubit properties. I have no good idea, but just want to bring it up.

What I was going for here was t1, t2, and maybe frequency (I'm not actually sure) would be generally applicable to any device that wanted to return qubit properties. So they got top level named attributes, but anything outside of that was a custom thing for a particular backend and would go in the `properties` dict. Part of what I wanted to try and do in this PR was to make it clearer for users and provider authors which fields were required, optional, or custom. In BackendV1 it's hard to tell the distinction between these 3 groups via `configuration()`, `properties()`, and `defaults()` because it just mirrors IBM's payload format and everything is basically flat. But, I think both you and @nkanazawa1989 are correct and this might be a bit weird from a user perspective. Do you think encouraging subclassing as @nkanazawa1989 suggested instead of having the `properties` field is sufficient to address this?

As for the qubit properties used by the transpiler, yeah ideally we'd put those in the target. I want the target to be the representation of a backend for the transpiler (as called by the preset pass managers). Nothing in `transpile()` today takes T1/T2 as a parameter that I'm aware of. But, I think we can expand on it over time in a non-breaking way if needed. So for example if in 0.20.0 we add a pass to the default pass managers that can use t1 we should expand the target to have an interface for it. It might be necessary to start versioning the target too to manage these changes over time which I think is fine (I didn't do it here for simplicity but nothing here precludes it) as I don't expect any of the interfaces here to be perfect or set in stone. Part of moving to a versioned backend interface is we can make iterative improvements and changes over time in a controlled manner as our needs change.

Reply:
Encouraging subclassing sounds good to me. I agree these properties can be provided if it is a qubit. Can you comment this on the class docs note?

Reply:
I added a note on this to the class documentation here: https://github.com/Qiskit/qiskit-terra/pull/5885/commits/2031f7fe1f0cce168ce8dfa03a577fb9c44dbe06

Reply:
So is `properties` needed then if they can subclass and add new ones? I think if IBM subclasses and adds anharmonicity, it should live at the same level as t1/t2/frequency, not one level lower (as a side note I think there might be a better name than `properties` since the others are also properties, but my preference is seeing if we can drop it altogether)

Reply:
Yeah, I think we can drop it as a field and just encourage subclassing. I'll remove it now

Reply:
Dropped in: https://github.com/Qiskit/qiskit-terra/pull/5885/commits/2942118b0d441757aaeef1333f72cbdf53e05452 it does mean we lose extra properties from the string representation of a `Target`, but that's probably ok